### PR TITLE
Add more array methods: `find_custom`, `rfind_custom`

### LINKED ISSUE #6889: Add more array methods: first_custom, last_custom, count_custom
### Describe the project you are working on

This is not unique to my project, but one project I am working on requires a lot of array traversals searching for elements that satisfy specific conditions. I feel there is a gap in Array functionality (because sort_custom already exists), and this would plug it,

### Describe the problem or limitation you are having in your project

There is no built-in methods to search for the first, last, and to count elements that satisfy custom conditions implemented in a Callable, the way you can filter or sort_custom.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add the following methods:
`find_custom(fn:Callable)` will return the index of the first element of an array where the callable fn will return true.
`rfind_custom(fn:Callable)` will return the index of the last element of an array where the callable fn will return true.
`count_custom(fn:Callable)` will return the count of all elements of the array where the callable fn will return true.

I attempted to name the functions to fit the naming style of the rest of the functions in the Array class. However, I do not like it- I would rather name it something like `find_by()`. But my gripes with the Array function naming conventions are beyond the scope of this proposal.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is a sketch implementation I have already created:
```cpp
int Array::find_custom(const Callable& p_callable, int p_from = 0) const {
	if (_p->array.size() == 0) {
		return -1;
	}

	if (p_from < 0 || size() == 0) {
		return -1;
	}

	const Variant* argptrs[1];

	for (int i = p_from; i < size(); i++) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(-1, "Error calling method from 'first_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Return as soon as possible
		if (result.operator bool()) {
			return i;
		}
	}

	return -1;
}

int Array::rfind_custom(const Callable& p_callable, int p_from = -1) const {
	if (_p->array.size() == 0) {
		return -1;
	}

	if (p_from < 0) {
		// Relative offset from the end
		p_from = _p->array.size() + p_from;
	}
	if (p_from < 0 || p_from >= _p->array.size()) {
		// Limit to array boundaries
		p_from = _p->array.size() - 1;
	}

	const Variant* argptrs[1];

	for (int i = p_from; i >= 0; i--) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(-1, "Error calling method from 'last_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Return as soon as possible
		if (result.operator bool()) {
			return i;
		}
	}

	return -1;
}

int Array::count_custom(const Callable& p_callable) const {
	int count = 0;
	const Variant* argptrs[1];

	for (int i = 0; i < size(); i++) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(0, "Error calling method from 'count_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Increase count if passes
		if (result.operator bool()) {
			count++;
		}
	}

	return count;
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, one can simply create a loop to accomplish the same task. But it must be done every time, and could be solved by patching this gap in GDScript.

### Is there a reason why this should be core and not an add-on in the asset library?

This adds core Array functionality.

### PR DESCRIPTION
**Pull request by SlashScreen**
Added 3 new array functions - `count_custom`, `first_custom`, and `last_custom`. These perform the functions of the existing `count,`, `first`, and `rfind` respectively, but using a Callable as the conditions, like `sort_custom`.
Last could be renamed `rfind_custom` for consistency sake.

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/6889*

### CODE REVIEW

**Comment by AThousandShips** [comment]
Please fix the name of your commit

**Comment by AThousandShips** [comment]
These need to be added to the binds for them to be useful in scripting

File: core/variant/array.cpp [hunk 1]
```diff
@@ -353,6 +353,36 @@ int Array::find(const Variant &p_value, int p_from) const {
 	return ret;
 }
 
+int Array::first_custom(const Callable& p_callable, int p_from = 0) const {
```
Comment (AThousandShips) [inline core/variant/array.cpp:356]: This name feels misleading, using `first` makes it sound like it returns that value, not the index

**Comment by SlashScreen** [comment]
You're right, I didn't think about using reduce in that way. I can remove that. Should I rename `last_custom` to `rfind_custom`?
As for the `first` ambiguity, that too is correct, but does saying `find_custom` make it sound like it does the same thing as `any`? Or no, because `any` returns a bool.
Also, yeah, not sure how I thought it would work without binding. I couldn't find where it was bound before, but after another look, I found it.

**Comment by dalexeev** [comment]
> Should I rename `last_custom` to `rfind_custom`?

I think it should be `find_custom` and `rfind_custom`, not `first_custom` and `last_custom`, for consistency with the already existing naming scheme.

**Comment by SlashScreen** [comment]
How would I create a callable for the tests? I don't see precedent for this, though I could be missing something.

**Comment by dalexeev** [comment]
> How would I create a callable for the tests?

I think you can use `callable_mp(&object, &Class::method)` or `callable_mp_static(&Class::static_method)`.

**Comment by AThousandShips** [comment]
Or add it as a gdscript test under `modules/gdscript/tests/runtime/features`

**Comment by SlashScreen** [comment]
> > How would I create a callable for the tests?
> 
> I think you can use `callable_mp(&object, &Class::method)`:
> 
> https://github.com/godotengine/godot/blob/237bd0a615df8a0e57bc3d299894abece7b43a0c/tests/core/object/test_object.h#L351
> 
> or `callable_mp_static(&Class::static_method)`:
> 
> https://github.com/godotengine/godot/blob/237bd0a615df8a0e57bc3d299894abece7b43a0c/tests/core/threads/test_worker_thread_pool.h#L46-L48
> 
> https://github.com/godotengine/godot/blob/237bd0a615df8a0e57bc3d299894abece7b43a0c/tests/core/threads/test_worker_thread_pool.h#L63

Ah, so I can use a c++ function as a callable?

**Comment by dalexeev** [comment]
> Ah, so I can use a c++ function as a callable?

Yes.

**Comment by SlashScreen** [comment]
I hope I amended it correctly. 
I added the test, removed count custom, added some documentation, and added some tests.

File: core/variant/variant_call.cpp [hunk 2]
```diff
@@ -2221,7 +2221,9 @@ static void _register_variant_builtin_methods() {
 	bind_method(Array, back, sarray(), varray());
 	bind_method(Array, pick_random, sarray(), varray());
 	bind_method(Array, find, sarray("what", "from"), varray(0));
+	bind_method(Array, find_custom, sarray("func", "from"), varray(0));
 	bind_method(Array, rfind, sarray("what", "from"), varray(-1));
+	bind_method(Array, frind_custom, sarray("func", "from"), varray(-1));
```
Comment (YuriSizov) [inline core/variant/variant_call.cpp:2226]: You have a typo here.

**Comment by YuriSizov** [comment]
Please also make sure that the commit message refers to the actual names of the method.

Personally, I agree that these should be variants of `find`. `first_custom` and `last_custom` don't make sense semantically. So the current state of this PR seems fine to me. Provided we agree to add these at all, of course.

**Comment by SlashScreen** [comment]
If there's any more changes that need be made, I will address them tomorrow, as it is 3:30 in the morning and my brain is shutting down. 
Could someone double-check that I created the callable properly in the tests?

File: core/variant/array.cpp [hunk 3]
```diff
@@ -378,6 +408,41 @@ int Array::rfind(const Variant &p_value, int p_from) const {
 	return -1;
 }
 
+int Array::rfind_custom(const Callable& p_callable, int p_from = -1) const {
+	if (_p->array.size() == 0) {
+		return -1;
+	}
+
+	if (p_from < 0) {
+		// Relative offset from the end
```
Comment (AThousandShips) [inline core/variant/array.cpp:417]:
```suggestion
		// Relative offset from the end.
```

File: core/variant/array.cpp [hunk 4]
```diff
@@ -378,6 +408,41 @@ int Array::rfind(const Variant &p_value, int p_from) const {
...
+		p_from = _p->array.size() + p_from;
+	}
+	if (p_from < 0 || p_from >= _p->array.size()) {
+		// Limit to array boundaries
```
Comment (AThousandShips) [inline core/variant/array.cpp:421]:
```suggestion
		// Limit to array boundaries.
```

File: core/variant/array.cpp [hunk 5]
```diff
@@ -378,6 +408,41 @@ int Array::rfind(const Variant &p_value, int p_from) const {
...
+		argptrs[0] = &get(i);
+
+		Variant result;
+		Callable::CallError ce;
+
+		p_callable.callp(argptrs, 1, result, ce);
+		if (ce.error != Callable::CallError::CALL_OK) {
+			ERR_FAIL_V_MSG(-1, "Error calling method from 'last_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
+		}
+		// Return as soon as possible
```
Comment (AThousandShips) [inline core/variant/array.cpp:437]:
```suggestion
		// Return as soon as possible.
```

File: tests/core/variant/test_array.h [hunk 6]
```diff
@@ -129,6 +129,22 @@ TEST_CASE("[Array] has() and count()") {
...
 }
 
+TEST_CASE("[Array] find_custom() and rfind_custom()") {
+	Array arr;
+	arr.push_back(1);
+	arr.push_back(2);
+	arr.push_back(3);
+	arr.push_back(4);
+
+	CHECK(arr.find_custom(callable_mp(&arr, array_test_callable)) == 1) // Should be the number 2
```
Comment (AThousandShips) [inline tests/core/variant/test_array.h:139]:
```suggestion
	CHECK(arr.find_custom(callable_mp_static(array_test_callable)) == 1) // Should be the number 2.
```
Fix for callable, and comment

File: tests/core/variant/test_array.h [hunk 7]
```diff
@@ -129,6 +129,22 @@ TEST_CASE("[Array] has() and count()") {
...
+	CHECK(arr.rfind_custom(callable_mp(&arr, array_test_callable)) == 3) // Should be the number 4
```
Comment (AThousandShips) [inline tests/core/variant/test_array.h:140]:
```suggestion
	CHECK(arr.rfind_custom(callable_mp_static(array_test_callable)) == 3) // Should be the number 4.
```

File: tests/core/variant/test_array.h [hunk 8]
```diff
@@ -129,6 +129,22 @@ TEST_CASE("[Array] has() and count()") {
...
+}
+
+static bool array_test_callable(*int val) {
+	// Basic even number check
```
Comment (AThousandShips) [inline tests/core/variant/test_array.h:144]:
```suggestion
	// Basic even number check.
```

File: tests/core/variant/test_array.h [hunk 9]
```diff
@@ -129,6 +129,22 @@ TEST_CASE("[Array] has() and count()") {
...
+	arr.push_back(1);
+	arr.push_back(2);
+	arr.push_back(3);
+	arr.push_back(4);
+
+	CHECK(arr.find_custom(callable_mp(&arr, array_test_callable)) == 1) // Should be the number 2
+	CHECK(arr.rfind_custom(callable_mp(&arr, array_test_callable)) == 3) // Should be the number 4
+}
+
+static bool array_test_callable(*int val) {
```
Comment (AThousandShips) [inline tests/core/variant/test_array.h:143]:
```suggestion
static bool array_test_callable(int val) {
```
Unsure what was the intended result here

File: core/variant/array.cpp [hunk 10]
```diff
@@ -378,6 +408,41 @@ int Array::rfind(const Variant &p_value, int p_from) const {
...
+	if (p_from < 0) {
+		// Relative offset from the end
+		p_from = _p->array.size() + p_from;
+	}
+	if (p_from < 0 || p_from >= _p->array.size()) {
+		// Limit to array boundaries
+		p_from = _p->array.size() - 1;
+	}
+
+	const Variant* argptrs[1];
```
Comment (AThousandShips) [inline core/variant/array.cpp:425]:
```suggestion
	const Variant *argptrs[1];
```

File: core/variant/array.cpp [hunk 11]
```diff
@@ -378,6 +408,41 @@ int Array::rfind(const Variant &p_value, int p_from) const {
 	return -1;
 }
 
+int Array::rfind_custom(const Callable& p_callable, int p_from = -1) const {
```
Comment (AThousandShips) [inline core/variant/array.cpp:411]:
```suggestion
int Array::rfind_custom(const Callable &p_callable, int p_from = -1) const {
```

File: core/variant/array.cpp [hunk 12]
```diff
@@ -353,6 +353,36 @@ int Array::find(const Variant &p_value, int p_from) const {
 	return ret;
 }
 
+int Array::find_custom(const Callable& p_callable, int p_from = 0) const {
```
Comment (AThousandShips) [inline core/variant/array.cpp:356]:
```suggestion
int Array::find_custom(const Callable &p_callable, int p_from = 0) const {
```

File: core/variant/array.cpp [hunk 13]
```diff
@@ -353,6 +353,36 @@ int Array::find(const Variant &p_value, int p_from) const {
...
+	if (_p->array.size() == 0) {
+		return -1;
+	}
+
+	if (p_from < 0 || size() == 0) {
+		return -1;
+	}
+
+	const Variant* argptrs[1];
```
Comment (AThousandShips) [inline core/variant/array.cpp:365]:
```suggestion
	const Variant *argptrs[1];
```

File: core/variant/array.h [hunk 14]
```diff
@@ -87,9 +87,11 @@ class Array {
 	int bsearch_custom(const Variant &p_value, const Callable &p_callable, bool p_before = true) const;
 	void reverse();
 
-	int find(const Variant &p_value, int p_from = 0) const;
-	int rfind(const Variant &p_value, int p_from = -1) const;
-	int count(const Variant &p_value) const;
+	int find(const Variant& p_value, int p_from = 0) const;
+	int find_custom(const Callable& p_callable, int p_from = 0) const;
```
Comment (AThousandShips) [inline core/variant/array.h:91]:
```suggestion
	int find_custom(const Callable &p_callable, int p_from = 0) const;
```

File: core/variant/array.h [hunk 15]
```diff
@@ -87,9 +87,11 @@ class Array {
...
+	int rfind(const Variant& p_value, int p_from = -1) const;
+	int rfind_custom(const Callable& p_callable, int p_from = -1) const;
```
Comment (AThousandShips) [inline core/variant/array.h:93]:
```suggestion
	int rfind_custom(const Callable &p_callable, int p_from = -1) const;
```

File: doc/classes/Array.xml [hunk 16]
```diff
@@ -560,6 +568,14 @@
 				Searches the array in reverse order. Optionally, a start search index can be passed. If negative, the start index is considered relative to the end of the array.
 			</description>
 		</method>
+		<method name="rfind_custom" qualifiers="const">
+			<return type="int" />
+			<param index="0" name="func" type="Callable" />
+			<param index="1" name="from" type="int" default="-1" />
+			<description>
+				Searches an array for the last value that causes [code]func[/code] to return true by searching in reverse order.  Optionally, a start search index can be passed. If negative, the start index is considered relative to the end of the array.
```
Comment (AThousandShips) [inline doc/classes/Array.xml:576]:
```suggestion
				Searches an array for the last value that causes [param func] to return true by searching in reverse order. Optionally, a start search index can be passed. If negative, the start index is considered relative to the end of the array.
```

File: doc/classes/Array.xml [hunk 17]
```diff
@@ -307,6 +307,14 @@
 				Searches the array for a value and returns its index or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
 			</description>
 		</method>
+		<method name="find_custom" qualifiers="const">
+			<return type="int" />
+			<param index="0" name="func" type="Callable" />
+			<param index="1" name="from" type="int" default="0" />
+			<description>
+				Searches an array for the first value that causes [code]func[/code] to return true, or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
```
Comment (AThousandShips) [inline doc/classes/Array.xml:315]:
```suggestion
				Searches an array for the first value that causes [param func] to return true, or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
```

**Comment by RedMser** [comment]
The method documentation should include that the `func` callable only gets called with one parameter: once for each element in the array. See the documentation of `map`, since I think it's a lot clearer.
E.g. you could otherwise assume it also includes the index as a second parameter or something (which I kinda wish for, but this needs a separate proposal and PR).

**Comment by SlashScreen** [comment]
Resolved typos and style checks from @AThousandShips .

**Comment by SlashScreen** [comment]
Fixed a bunch of small mistakes picked up by the compile check.

File: tests/core/variant/test_array.h [hunk 18]
```diff
@@ -129,6 +129,22 @@ TEST_CASE("[Array] has() and count()") {
...
+
+TEST_CASE("[Array] find_custom() and rfind_custom()") {
+	Array arr;
+	arr.push_back(1);
+	arr.push_back(2);
+	arr.push_back(3);
+	arr.push_back(4);
+
+	CHECK(arr.find_custom(callable_mp_static(array_test_callable)) == 1); // Should be the number 2.
+	CHECK(arr.rfind_custom(callable_mp_static(array_test_callable)) == 3); // Should be the number 4.
```
Comment (dalexeev) [inline tests/core/variant/test_array.h:145]: There are no tests with `from` argument.

Comment (SlashScreen) [reply tests/core/variant/test_array.h:145]: Good point. I'll add that shortly.

**Comment by SlashScreen** [comment]
Whoops, I messed up the test. That's what I get for having nvim in a thin window where I can't see the whole line.

**Comment by YuriSizov** [comment]
You still need to amend the commit message to reflect what the methods are actually called.

**Comment by SlashScreen** [comment]
nice catch.

File: doc/classes/Array.xml [hunk 19]
```diff
@@ -307,6 +307,15 @@
 				Searches the array for a value and returns its index or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
 			</description>
 		</method>
+		<method name="find_custom" qualifiers="const">
+			<return type="int" />
+			<param index="0" name="func" type="Callable" />
+			<param index="1" name="from" type="int" default="0" />
+			<description>
+				Searches an array for the first value that causes [param func] to return true, or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
+				[param func] is a callable that should accept an element array as an argument, and return a boolean.
```
Comment (AThousandShips) [inline doc/classes/Array.xml:316]:
```suggestion
				[param func] is a callable that should accept an array element as an argument, and return a boolean.
```
The passed value is an element from the array, same for the other method

File: doc/classes/Array.xml [hunk 20]
```diff
@@ -307,6 +307,15 @@
 				Searches the array for a value and returns its index or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
 			</description>
 		</method>
+		<method name="find_custom" qualifiers="const">
+			<return type="int" />
+			<param index="0" name="func" type="Callable" />
+			<param index="1" name="from" type="int" default="0" />
+			<description>
+				Searches an array for the first value that causes [param func] to return true, or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
```
Comment (AThousandShips) [inline doc/classes/Array.xml:315]:
```suggestion
				Searches the array for the first value that causes [param func] to return true, or [code]-1[/code] if not found. Optionally, the initial search index can be passed.
```
To match other cases, and as this is a member function it makes sense to refer to this, same for other

File: core/variant/array.cpp [hunk 21]
```diff
@@ -353,6 +353,36 @@ int Array::find(const Variant &p_value, int p_from) const {
...
+
+	for (int i = p_from; i < size(); i++) {
+		argptrs[0] = &get(i);
+
+		Variant result;
+		Callable::CallError ce;
+
+		p_callable.callp(argptrs, 1, result, ce);
+		if (ce.error != Callable::CallError::CALL_OK) {
+			ERR_FAIL_V_MSG(-1, "Error calling method from 'find_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
```
Comment (AThousandShips) [inline core/variant/array.cpp:375]:
```suggestion
			ERR_FAIL_V_MSG(-1, "Error calling method from 'find_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce) + ".");
```

File: core/variant/array.cpp [hunk 22]
```diff
@@ -378,6 +408,41 @@ int Array::rfind(const Variant &p_value, int p_from) const {
...
+
+	for (int i = p_from; i >= 0; i--) {
+		argptrs[0] = &get(i);
+
+		Variant result;
+		Callable::CallError ce;
+
+		p_callable.callp(argptrs, 1, result, ce);
+		if (ce.error != Callable::CallError::CALL_OK) {
+			ERR_FAIL_V_MSG(-1, "Error calling method from 'rfind_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
```
Comment (AThousandShips) [inline core/variant/array.cpp:435]:
```suggestion
			ERR_FAIL_V_MSG(-1, "Error calling method from 'rfind_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce) + ".");
```

**Review by AThousandShips — APPROVED** [review]
Looks good save for some corrections to the documentation and style

**Comment by SlashScreen** [comment]
Integrated the changes, @AThousandShips .

File: doc/classes/Array.xml (hunk 19 above)
Comment (AThousandShips) [reply doc/classes/Array.xml]: This and the next need to be applied for `rfind_custom` as well

**Comment by AThousandShips** [comment]
Turns out there's already a PR for this change, predating it:
* https://github.com/godotengine/godot/pull/73490

But this one is more refined IMO

**Comment by SlashScreen** [comment]
oh dear, I'm not sure i updated the documentation for the other function... Will the checks need to be run again?

**Comment by AThousandShips** [comment]
They do, but no worries just go ahead and fix them

**Comment by maximkulkin** [comment]
This one is exactly same as #73490 minus naming with an exception that the other one has much more comprehensive test suite.

**Comment by SlashScreen** [comment]
So is mine redundant?

**Comment by AThousandShips** [comment]
Will be a decision which one to merge

**Comment by SlashScreen** [comment]
Changed semantics on rfind_custom().

**Comment by AThousandShips** [comment]
Reviewers, see also:
* https://github.com/godotengine/godot/pull/70727
* https://github.com/godotengine/godot/pull/73490

**Comment by Calinou** [comment]
- Superseded by https://github.com/godotengine/godot/pull/95449.

**Pull request closed without merging**